linkerd/app/gateway/src/lib.rs

```diff
@@ -0,0 +1,184 @@
+    #[tokio::test]
+    async fn forward_loop() {
+        let test = Test {
+            orig_fwd: Some(
+                "by=gateway.id.test;for=client.id.test;host=dst.test.example.com:4321;proto=https",
+            ),
+            ..Default::default()
+        };
+        let status = test
+            .run()
+            .await
+            .unwrap_err()
+            .downcast_ref::<HttpError>()
+            .unwrap()
+            .status();
+        assert_eq!(status, http::StatusCode::LOOP_DETECTED);
+    }
+
+    struct Test {
+        suffix: &'static str,
+        dst_name: Option<&'static str>,
+        peer_id: tls::PeerIdentity,
+        orig_fwd: Option<&'static str>,
+    }
+
+    impl Default for Test {
+        fn default() -> Self {
+            Self {
+                suffix: "test.example.com",
+                dst_name: Some("dst:4321"),
+                peer_id: tls::PeerIdentity::Some(identity::Name::from(
+                    dns::Name::try_from("client.id.test".as_bytes()).unwrap(),
+                )),
+                orig_fwd: None,
+            }
+        }
+    }
+
+    impl Test {
+        async fn run(self) -> Result<http::Response<http::boxed::Payload>, Error> {
+            let Self {
+                suffix,
+                dst_name,
+                peer_id,
+                orig_fwd,
+            } = self;
+
+            let (outbound, mut handle) = mock::pair::<
+                http::Request<http::boxed::Payload>,
+                http::Response<http::boxed::Payload>,
+            >();
+            let mut make_gateway = {
+                let resolve = service_fn(move |short: dns::Name| async move {
+                    let long = format!("{}.{}", short.without_trailing_dot(), suffix);
+                    Ok::<_, Never>((
+                        dns::Name::try_from(long.as_bytes()).unwrap(),
+                        IpAddr::from([127, 0, 0, 1]),
+                    ))
+                });
+                mock::Spawn::new(make::MakeGateway::new(
+                    resolve,
+                    service_fn(move |_: _| {
+                        let out = outbound.clone();
+                        async move { Ok::<_, Never>(out) }
+                    }),
+                    tls::PeerIdentity::Some(identity::Name::from(
+                        dns::Name::try_from("gateway.id.test".as_bytes()).unwrap(),
+                    )),
+                    Some(dns::Suffix::try_from("test.example.com.").unwrap()),
+                ))
+            };
+
+            let target = inbound::Target {
+                addr: SocketAddr::from(([127, 0, 0, 1], 4143)),
+                dst_name: dst_name.map(|n| NameAddr::from_str(n).unwrap()),
+                http_settings: http::Settings::Http2,
+                tls_client_id: peer_id,
+            };
+            assert_ready_ok!(make_gateway.poll_ready());
+            let gateway = make_gateway.call(target).await.unwrap();
+
+            tokio::spawn(async move {
```

Comment:
if the spawned task panics, we should probably fail the test, right? might be good to hang onto its `JoinHandle` and await & unwrap it after we await the gateway request, just to make sure.